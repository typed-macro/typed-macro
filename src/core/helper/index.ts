export { BABEL_TOOLS, Babel } from './babel'
export { ImportOption } from './import'
export { getPathHelper, PathHelper } from './path'
export { getStateHelper, StateHelper } from './state'
export { getTransformHelper, TransformHelper } from './transform'
