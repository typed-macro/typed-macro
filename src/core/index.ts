export { Runtime, RuntimeOptions } from './runtime'
export {
  createTransformer,
  Transformer,
  TransformerOptions,
} from './transformer'
export {
  Macro,
  MacroContext,
  MacroHelper,
  MacroHandler,
  Exportable,
  NamespacedExportable,
  NormalizedExports,
  normalizeExports,
} from './types'
