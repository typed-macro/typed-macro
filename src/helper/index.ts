export * from './macro'
export * from './babel'
export type { ImportOption } from './import'
export type { DevServerHelper } from './server'
