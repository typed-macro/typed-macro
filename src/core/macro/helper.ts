import { ImportOption } from '@/core/helper/import'
import { Node, NodePath } from '@babel/traverse'
import { CallExpression, ImportDeclaration, Program } from '@babel/types'
import {
  normalizePathPattern,
  PathPatternNormalizer,
  projectDir,
} from '@/core/helper/path'
import {
  containsMacros,
  findImported,
  ImportedMacro,
} from '@/core/helper/traverse'
import {
  appendImports,
  appendToBody,
  prependImports,
  prependToBody,
} from '@/core/helper/transform'
import { promise } from '@/common'

export type MacroHelper = {
  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement,
   *   or undefined if imports param is array and length === 0
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   *   or undefined if imports param is array and length === 0
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Normalize path pattern so can resolve file paths as import paths
   * @param path relative path or path pattern
   * @param root absolute file path of the project dir, default to projectDir('leaf')
   * @param importer absolute file path of the importer, default to current file
   *
   * e.g. search files by glob pattern and then import them
   *
   * ```typescript
   * // /src/example/a.ts
   * // /src/example/b.ts
   * // in /src/another/index.ts
   * const pattern = '../example/*.ts'
   * const { normalized, base, resolveImportPath } = normalizePathPattern(pattern)
   * const importPaths = glob.sync(normalized, { cwd: base }).map(resolveImportPath)
   *
   * // importPaths = ['../example/a.ts', '../example/b.ts']
   * ```
   */
  normalizePathPattern: (
    pattern: string,
    root?: string,
    importer?: string
  ) => PathPatternNormalizer

  /**
   * Get the directory of the root project or the nearest project.
   * @param which 'root' or 'leaf'
   */
  projectDir: (which: 'root' | 'leaf') => string

  /**
   * Check whether there are unexpanded macros under the path(s).
   * @param path the node path(s) to be checked, use the arguments of the current
   * call expression by default
   */
  containsMacros: (...paths: NodePath[]) => boolean[]

  /**
   * Yield current transformer to nested macros inside current
   * call expression and then call the current macro handler again
   * if the nested macros exist.
   * In another words, expand nested macros in arguments first.
   *
   * It is recommended to call this function at the beginning of the handler
   * that needs to ensure that the arguments don't contain nested macros,
   * to reduce unnecessary repeated calculations.
   *
   * e.g.
   * ```typescript
   * ({ path }, _, { yieldToNestedMacros }) => {
   *   yieldToNestedMacros()
   *   // ...your code
   * }
   * ```
   *
   * Internally, in v0.2.x, it throws a Promise, so DO NOT call it in a try block.
   */
  yieldToNestedMacros: () => never | void
}

export function createHelper(
  thisPath: NodePath<CallExpression>,
  thisProgram: NodePath<Program>,
  thisFilepath: string,
  importedMacros: ImportedMacro[]
) {
  const helpers: MacroHelper = {
    findImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose)
    },

    hasImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose) !== undefined
    },

    prependImports: (imports, program = thisProgram) => {
      return prependImports(program, imports)
    },

    appendImports: (imports, program = thisProgram) => {
      return appendImports(program, imports)
    },

    prependToBody: (nodes, program = thisProgram) => {
      return prependToBody(program, nodes)
    },

    appendToBody: (nodes, program = thisProgram) => {
      return appendToBody(program, nodes)
    },

    normalizePathPattern: (
      pattern,
      root = projectDir('leaf'),
      importer = thisFilepath
    ) => {
      return normalizePathPattern(pattern, importer, root)
    },

    projectDir,

    containsMacros: (...paths) => {
      return containsMacros(
        paths.length ? paths : thisPath.get('arguments'),
        importedMacros
      )
    },

    yieldToNestedMacros: () => {
      if (
        containsMacros(thisPath.get('arguments'), importedMacros).includes(true)
      )
        throw promise
    },
  }

  return helpers
}
