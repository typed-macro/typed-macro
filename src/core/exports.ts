import { isMacro, Macro } from '@/core/macro'
import { isMacroCompatible } from '@/core/compat'

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string[]
    macroScope: string[]
  }
}

export type Exportable = (
  | {
      /**
       * as macros
       */
      macros: Macro[]
    }
  | {
      /**
       * as module, code will be used as source code of the module
       */
      code: string
    }
) & {
  /**
   * Type definitions, will be written to d.ts.
   *
   * e.g.
   * ```typescript
   * { exports: { '@macros': { customTypes: `type A = string` } } }
   * ```
   * will generate
   * ```typescript
   * declare module '@macros' {
   *   type A = string
   * }
   * ```
   */
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes ? [item.customTypes] : [],
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacro(m)) throw new Error(`'${m.name}' is not a macro`)
        return m.__types
      })
    }
  })

  validateMacros(macros)

  return {
    macros,
    modules,
    types,
  }
}

// ensure compatible and no name conflict
export function validateMacros(macros: NamespacedMacros) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (!isMacroCompatible(m))
        throw new Error(`macro '${m.name}' in '${ns}' is incompatible`)

      if (mem[m.name])
        throw new Error(`macro '${m.name}' in '${ns}' is duplicated`)

      mem[m.name] = 1
    })
  })
}
