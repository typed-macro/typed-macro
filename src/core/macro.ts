import { CallExpression } from '@babel/types'
import { NodePath } from '@babel/traverse'
import { Babel, PathHelper, StateHelper, TransformHelper } from '@/core/helper'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /*
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath
  /*
   * Whether in ssr mode.
   */
  ssr: boolean
}

export type MacroHelper = TransformHelper & PathHelper & StateHelper

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

export type Macro = {
  name: string
  apply: MacroHandler
}

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  return {
    name,
    apply: handler,
    __types: renderMetaType(name, meta),
  } as InternalMacro
}

export function isMacro(o: unknown): o is InternalMacro {
  return !!(o as InternalMacro).__types
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
