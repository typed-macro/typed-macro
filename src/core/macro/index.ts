import { CallExpression, File, Program } from '@babel/types'
import { NodePath } from '@babel/traverse'

import { CURRENT_CALL_VERSION, Versioned } from '@/core/version'
import { State } from '../helper/state'
import { HELPER, MacroHelper } from './helper'
import { Babel, BABEL_TOOLS } from './babel'
import { ImportedMacro } from '@/core/helper/traverse'

export type MacroContext = {
  /**
   * The file path of the module currently being handled.
   */
  filepath: string
  /**
   * The original source code of the module currently being handled.
   */
  code: string
  /**
   * The arguments nodes of the call-macro expression currently being handled.
   */
  args: CallExpression['arguments']
  /**
   * The NodePath of the call-macro expression currently being handled.
   */
  path: NodePath<CallExpression>
  /**
   * The AST being traversed.
   */
  ast: File
  /**
   * Whether in ssr mode.
   */
  ssr: boolean
  /**
   * Traversal state.
   */
  state: {
    /**
     * In order to recursively expand all macros in a file,
     * the transformer will traverse the AST many times until all macros
     * are expanded.
     *
     * - {@link traversal} is the state shared during one traversal, and
     * - {@link transform} is the state shared during the whole transformation.
     */
    traversal: State
    /**
     * @see traversal
     */
    transform: State
  }
}

export type MacroHandler = (
  /**
   * Context about the macro caller.
   */
  ctx: MacroContext,
  /**
   * Babel tools to help operate node paths and nodes.
   */
  babel: Readonly<Babel>,
  /**
   * Wrappers upon babel tools to simplify operations.
   */
  helper: Readonly<MacroHelper>
) => void

type RawMacroCall = Versioned<{
  filepath: string
  code: string
  path: NodePath<CallExpression>
  ast: File
  program: NodePath<Program>
  ssr: boolean
  traversalState: State
  transformState: State
  importedMacros: ImportedMacro[]
}>

export type Macro = (call: RawMacroCall) => void

export type MacroMeta = {
  signatures: {
    comment?: string
    signature: string
  }[]
  types: string[]
}

type InternalMacro = Macro & {
  __types: string
}

export function macro(
  name: string,
  meta: MacroMeta,
  handler: MacroHandler
): Macro {
  const normalizer = getNormalizer(handler)
  const m = {
    [name](call: RawMacroCall) {
      const { ctx, babel, helper } = normalizer(call)
      handler(ctx, babel as Babel, helper as MacroHelper)
    },
  }[name]
  ;(m as InternalMacro).__types = renderMetaType(name, meta)
  return m
}

type MacroCallNormalizer = (raw: RawMacroCall) => {
  ctx: MacroContext
  babel?: Readonly<Babel>
  helper?: Readonly<MacroHelper>
}

function ensureCompatible(raw: RawMacroCall) {
  if (raw.__version !== CURRENT_CALL_VERSION)
    throw new Error('the macro is incompatible with the runtime')
}

function normalizeContext(raw: RawMacroCall): MacroContext {
  const { filepath, ssr, ast, code, path, transformState, traversalState } =
    raw.value
  return {
    filepath,
    code,
    ast,
    path,
    args: path.node.arguments,
    state: {
      transform: transformState,
      traversal: traversalState,
    },
    ssr,
  }
}

export function getNormalizer(handler: MacroHandler): MacroCallNormalizer {
  switch (handler.length) {
    case 0:
      throw new Error('not a valid macro handler')
    case 1:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        return {
          ctx: normalizeContext(raw),
        }
      }

    case 2:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        return {
          ctx: normalizeContext(raw),
          babel: BABEL_TOOLS,
        }
      }

    default:
      return (raw: RawMacroCall) => {
        ensureCompatible(raw)
        const ctx = normalizeContext(raw)
        const { path, filepath, program, importedMacros } = raw.value
        HELPER.props
          .setImportedMacros(importedMacros)
          .setPath(path)
          .setFilepath(filepath)
          .setProgram(program)
        return {
          ctx,
          babel: BABEL_TOOLS,
          helper: HELPER.instance,
        }
      }
  }
}

export function isMacro(o: unknown): o is InternalMacro {
  return (o as InternalMacro)?.__types !== undefined ?? false
}

export function renderMetaType(name: string, meta: MacroMeta) {
  return [
    meta.types.join('\n'),
    meta.signatures
      .map((s) =>
        s.comment
          ? `  /** ${s.comment} **/
  export function ${name}${s.signature}`
          : `  export function ${name}${s.signature}`
      )
      .join('\n'),
  ]
    .filter((t) => !!t)
    .join('\n')
}
