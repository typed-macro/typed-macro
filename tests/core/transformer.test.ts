import { getAST, getExpression, matchCodeSnapshot, NO_OP } from '../testutils'
import { CallExpression, StringLiteral } from '@babel/types'
import {
  applyMacros,
  collectImportedMacros,
  createTransformer,
  findCalledMacro,
} from '@/core/transformer'
import { Macro } from '@/core/macro'
import { NamespacedMacros } from '@/core/exports'
import { getStateHelper } from '@/core/helper'

describe('collectImportedMacros()', () => {
  it('should work', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      { code: `import { a } from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
      },
      { code: `import a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import * as a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import { b } from '@b'`, macros: { '@a': [{ name: 'a' }] } },
    ]
    testCases.forEach((c) => {
      expect(
        collectImportedMacros(getAST(c.code), c.macros as any)
      ).toMatchSnapshot()
    })
  })

  it('should rewrite import stmts if keep=true', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      { code: `import { a } from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
      },
      { code: `import a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import * as a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      collectImportedMacros(ast, c.macros as any, true)
      matchCodeSnapshot(ast)
    })
  })

  it('should remove import stmts if keep=false', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      { code: `import { a } from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
      },
      { code: `import a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
      { code: `import * as a from '@a'`, macros: { '@a': [{ name: 'a' }] } },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      collectImportedMacros(ast, c.macros as any)
      expect(ast.program.body.length).toBe(0)
    })
  })
})

describe('findCalledMacro()', () => {
  it('should work', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
      call: string
    }[] = [
      {
        code: `import { a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import { a as _a } from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `_a()`,
      },
      {
        code: `import a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a.a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'a' }] },
        call: `a['a']()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'b' }] },
        call: `a.a()`,
      },
      {
        code: `import * as a from '@a'`,
        macros: { '@a': [{ name: 'b' }] },
        call: `a['a']()`,
      },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      const importedMacros = collectImportedMacros(ast, c.macros as any)

      expect(
        findCalledMacro(
          (getExpression(c.call) as CallExpression).callee,
          importedMacros
        )
      ).toMatchSnapshot()
    })
  })
})

describe('applyMacros()', () => {
  it('should work', () => {
    const testCases: {
      code: string
      macros: Record<string, Macro[]>
    }[] = [
      {
        code: `import {a} from '@a'`,
        macros: {
          '@a': [
            {
              name: 'a',
              apply: jest.fn(),
            },
          ],
        },
      },
      {
        code: `import {a} from '@a'; a()`,
        macros: {
          '@a': [
            {
              name: 'a',
              apply: jest.fn(),
            },
          ],
        },
      },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      const importedMacros = collectImportedMacros(ast, c.macros)

      expect(
        applyMacros({
          code: c.code,
          filepath: '',
          ast,
          importedMacros,
          ssr: false,
          state: getStateHelper(),
        })
      ).toMatchSnapshot()
    })
  })

  it('should throw error when call non-existent macro', () => {
    const testCases: {
      code: string
      macros: Record<string, { name: string }[]>
    }[] = [
      {
        code: `import a from '@a'; a.a()`,
        macros: {
          '@a': [],
        },
      },
    ]
    testCases.forEach((c) => {
      const ast = getAST(c.code)
      const importedMacros = collectImportedMacros(ast, c.macros as any)

      expect(() =>
        applyMacros({
          code: c.code,
          filepath: '',
          ast,
          importedMacros,
          ssr: false,
          state: getStateHelper(),
        })
      ).toThrow()
    })
  })
})

describe('transformer', () => {
  it('should work', () => {
    const transform = createTransformer({
      parserPlugins: [],
      maxRecursion: 5,
    })

    expect(
      transform(
        {
          code: `import { echo } from '@echo'; echo('yeah')`,
          filepath: '',
          dev: true,
        },
        {
          '@echo': [
            {
              name: 'echo',
              apply: ({ path, args }, { template }) => {
                const msg = (args[0] as StringLiteral).value
                path.replaceWith(
                  template.statement.ast`console.log("${Array.from(
                    { length: 3 },
                    () => msg
                  ).join(' ')}")`
                )
              },
            },
          ],
        }
      )
    ).toMatchSnapshot()
  })

  it('should throw errors if there is wrong in steps', () => {
    const macros: NamespacedMacros = {
      '@echo': [
        {
          name: 'echo',
          apply: ({ path, args }) => {
            if (args.length === 0) throw new Error('')
            path.remove()
          },
        },
      ],
    }
    const transformer = createTransformer({
      parserPlugins: [],
      maxRecursion: 5,
    })
    const transform = (code: string) =>
      transformer(
        {
          code,
          filepath: '',
          dev: false,
        },
        macros
      )
    // args.length === 0
    expect(() => transform(`import { echo } from '@echo'; echo()`)).toThrow()
    // import non-existent macro and call
    expect(() => transform(`import { abc } from '@echo'; abc()`)).toThrow()
    // import non-existent macro but no call
    expect(() => transform(`import { abc } from '@echo';`)).not.toThrow()
    // call non-existent macro
    expect(() => transform(`import echo from '@echo'; echo.abc()`)).toThrow()
  })

  it('should throw errors if reached max recursion', () => {
    const transform = createTransformer({
      parserPlugins: [],
      maxRecursion: 5,
    })
    // recursion
    expect(() =>
      transform(
        {
          code: `import {echo} from '@echo'; echo('yeah')`,
          filepath: '',
          dev: false,
        },
        {
          '@echo': [
            {
              name: 'echo',
              apply: NO_OP,
            },
          ],
        }
      )
    ).toThrow()
  })

  it('should recollect imported macros every loop', () => {
    const fn = jest.fn()
    const transform = createTransformer({
      parserPlugins: [],
      maxRecursion: 5,
    })

    expect(
      transform(
        {
          code: `import { a } from '@echo'; a()`,
          filepath: '',
          dev: false,
        },
        {
          '@echo': [
            {
              name: 'a',
              apply: ({ path }, { template }, { appendToBody }) => {
                appendToBody(
                  template.statements.ast(`import {b} from '@echo'; b()`)
                )
                path.remove()
              },
            },
            {
              name: 'b',
              apply: ({ path }, { template }) => {
                fn()
                path.replaceWith(
                  template.statement.ast(`console.log('hello world')`)
                )
              },
            },
          ],
        }
      )
    ).toMatchSnapshot()
    expect(fn).toBeCalled()
  })
})
