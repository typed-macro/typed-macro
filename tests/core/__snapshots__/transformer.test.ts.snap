// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyMacros() should work 1`] = `false`;

exports[`applyMacros() should work 2`] = `true`;

exports[`collectImportedMacros() should rewrite import stmts if keep=true 1`] = `"import '@a';"`;

exports[`collectImportedMacros() should rewrite import stmts if keep=true 2`] = `"import '@a';"`;

exports[`collectImportedMacros() should rewrite import stmts if keep=true 3`] = `"import '@a';"`;

exports[`collectImportedMacros() should rewrite import stmts if keep=true 4`] = `"import '@a';"`;

exports[`collectImportedMacros() should work 1`] = `
Array [
  Object {
    "importedAsNamespace": false,
    "local": "a",
    "macro": Object {
      "name": "a",
    },
    "namespace": "@a",
  },
]
`;

exports[`collectImportedMacros() should work 2`] = `
Array [
  Object {
    "importedAsNamespace": false,
    "local": "_a",
    "macro": Object {
      "name": "a",
    },
    "namespace": "@a",
  },
]
`;

exports[`collectImportedMacros() should work 3`] = `
Array [
  Object {
    "importedAsNamespace": true,
    "local": "a",
    "macros": Array [
      Object {
        "name": "a",
      },
    ],
    "namespace": "@a",
  },
]
`;

exports[`collectImportedMacros() should work 4`] = `
Array [
  Object {
    "importedAsNamespace": true,
    "local": "a",
    "macros": Array [
      Object {
        "name": "a",
      },
    ],
    "namespace": "@a",
  },
]
`;

exports[`collectImportedMacros() should work 5`] = `Array []`;

exports[`findCalledMacro() should work 1`] = `
Object {
  "name": "a",
}
`;

exports[`findCalledMacro() should work 2`] = `undefined`;

exports[`findCalledMacro() should work 3`] = `
Object {
  "name": "a",
}
`;

exports[`findCalledMacro() should work 4`] = `undefined`;

exports[`findCalledMacro() should work 5`] = `undefined`;

exports[`findCalledMacro() should work 6`] = `
Object {
  "name": "a",
}
`;

exports[`findCalledMacro() should work 7`] = `
Object {
  "name": "a",
}
`;

exports[`findCalledMacro() should work 8`] = `"a"`;

exports[`findCalledMacro() should work 9`] = `"a"`;

exports[`transformer should recollect imported macros every loop 1`] = `"console.log('hello world');"`;

exports[`transformer should work 1`] = `"import '@echo';console.log(\\"yeah yeah yeah\\");"`;
